datafusion-federation/src/analyzer/mod.rs

```diff
@@ -112,39 +132,36 @@ impl FederationAnalyzerRule {
     }
 
     /// scans an expression to see if it belongs to a single [`FederationProvider`]
-    fn scan_expr_recursively(&self, expr: &Expr) -> Result<ScanResult> {
+    fn scan_expr_recursively(
+        &self,
+        expr: &Expr,
+        providers: &HashMap<TableReference, Arc<dyn FederationProvider>>,
+    ) -> Result<ScanResult> {
         let mut sole_provider: ScanResult = ScanResult::None;
 
         expr.apply(&mut |e: &Expr| -> Result<TreeNodeRecursion> {
-            // TODO: Support other types of sub-queries
             match e {
                 Expr::ScalarSubquery(ref subquery) => {
-                    let plan_result = self.scan_plan_recursively(&subquery.subquery)?;
+                    let plan_result = self.scan_plan_recursively(&subquery.subquery, providers)?;
 
                     sole_provider.merge(plan_result);
                     Ok(sole_provider.check_recursion())
                 }
                 Expr::InSubquery(ref insubquery) => {
-                    let plan_result = self.scan_plan_recursively(&insubquery.subquery.subquery)?;
+                    let plan_result =
+                        self.scan_plan_recursively(&insubquery.subquery.subquery, providers)?;
 
                     sole_provider.merge(plan_result);
                     Ok(sole_provider.check_recursion())
                 }
                 Expr::OuterReferenceColumn(_, ref col) => {
                     if let Some(table) = &col.relation {
-                        let map = self.provider_map.read().map_err(|_| {
-                            DataFusionError::External(
-                                "Failed to create federated plan: failed to obtain a read lock on federated providers.".into(),
-                            )
-                        })?;
-                        if let Some(plan_result) = map.get(table) {
-                            sole_provider.merge(plan_result.clone());
+                        if let Some(plan_result) = providers.get(table) {
```

Comment:
Instead of storing a ScanResult, we store the federation provider directly. It didn't really make sense to store the scan result previously.